src/Trial.hs

```diff
@@ -60,45 +62,84 @@ instance Semigroup (Trial e a) where
     Fiasco e1   <> Result e2 a = Result (map snd e1 <> e2) a
     Result e1 a <> Fiasco e2   = Result (e1 <> map snd e2) a
     Result e1 _ <> Result e2 b = Result (e1 <> e2) b
+    {-# INLINE (<>) #-}
+
+    sconcat :: NonEmpty (Trial e a) -> Trial e a
+    sconcat (x :| xs) = foldl' (<>) x xs
+    {-# INLINE sconcat #-}
 
 instance Functor (Trial e) where
     fmap :: (a -> b) -> Trial e a -> Trial e b
     fmap _ (Fiasco e)   = Fiasco e
     fmap f (Result e a) = Result e $ f a
+    {-# INLINE fmap #-}
+
+    (<$) :: a -> Trial e b -> Trial e a
+    _ <$ Fiasco e   = Fiasco e
+    a <$ Result e _ = Result e a
+    {-# INLINE (<$) #-}
 
 instance Applicative (Trial e) where
     pure :: a -> Trial e a
     pure = Result []
+    {-# INLINE pure #-}
 
     (<*>) :: Trial e (a -> b) -> Trial e a -> Trial e b
-    Fiasco e1 <*> Fiasco e2   = Fiasco $ e1 <> e2
-    Fiasco e1 <*> Result e2 _ = Fiasco $ e1 <> map (W,) e2
-    Result e1 _ <*> Fiasco e2 = Fiasco $ map (W,) e1 <> e2
-    Result e1 f <*> Result e2 a = Result (e1 <> e2) $ f a
+    Fiasco e1   <*> Fiasco e2   = Fiasco (e1 <> e2)
+    Fiasco e1   <*> Result e2 _ = Fiasco (e1 <> withW e2)
```

Comment:
Do you think that it would make sense to improve the result finding a little bit?
```suggestion
    Fiasco e1   <*> trial   = Fiasco $ case trial of
        Fiasco e2   -> e1 <> e2
        Result e2 _ -> e1 <> withW e2
```